rust-runtime/aws-smithy-runtime-api/src/interceptors.rs

```diff
@@ -565,31 +608,59 @@ impl<ModReq, TxReq, TxRes, ModRes> Interceptors<ModReq, TxReq, TxRes, ModRes> {
         Self::default()
     }
 
-    // pub fn with_interceptor(
-    //     &mut self,
-    //     interceptor: impl Interceptor<ModReq, TxReq, TxRes, ModRes>,
-    // ) -> &mut Self {
-    //     self.inner.push(Box::new(interceptor));
-    //
-    //     self
-    // }
+    pub fn with_client_interceptor(
+        &mut self,
+        interceptor: impl Interceptor<ModReq, TxReq, TxRes, ModRes> + 'static,
+    ) -> &mut Self {
+        self.client_interceptors.push(Box::new(interceptor));
+        self
+    }
+
+    pub fn with_operation_interceptor(
+        &mut self,
+        interceptor: impl Interceptor<ModReq, TxReq, TxRes, ModRes> + 'static,
+    ) -> &mut Self {
+        self.operation_interceptors.push(Box::new(interceptor));
+        self
+    }
+
+    fn all_interceptors_mut(
+        &mut self,
+    ) -> impl Iterator<Item = &mut Box<dyn Interceptor<ModReq, TxReq, TxRes, ModRes>>> {
+        self.client_interceptors
+            .iter_mut()
+            .chain(self.operation_interceptors.iter_mut())
+    }
+
+    pub fn client_read_before_execution(
+        &mut self,
+        context: &InterceptorContext<ModReq, TxReq, TxRes, ModRes>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), InterceptorError> {
+        for interceptor in self.client_interceptors.iter_mut() {
+            interceptor.read_before_execution(context, cfg)?;
+        }
+        Ok(())
+    }
 
-    pub fn read_before_execution(
+    pub fn operation_read_before_execution(
```

Comment:
Nit: I think we should make a macro to generate all these functions since they're completely identical except for the function names...